Cooperative emission failure in ReadMapP

When `readMapP` has a projection to an object that doesn't implement `equals`, the cooperative emission can fail with 
```
java.lang.AssertionError: Different item offered after previous call returned false: expected=[Ljava.lang.Object;@71e5cb0a, got=[Ljava.lang.Object;@75da41ee
```
I think it works correctly if assertions are disabled, though I did not test.

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/test/TestSupport.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/test/TestSupport.java
@@ -588,7 +588,7 @@ public final class TestSupport {
                 inboxOrdinal = input.get(inputPosition).ordinal;
                 for (int added = 0;
                         inputPosition < input.size()
-                                && added < testMode.inboxSize()
+                                && added < testMode.inboxLimit()
                                 && inboxOrdinal == input.get(inputPosition).ordinal
                                 && (added == 0 || !(input.get(inputPosition).item instanceof Watermark));
                         added++
@@ -983,7 +983,7 @@ public final class TestSupport {
         /**
          * Size limit of the inbox.
          */
-        public int inboxSize() {
+        public int inboxLimit() {
             return inboxLimit;
         }
 

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java
@@ -113,6 +113,8 @@ public final class ReadMapOrCacheP<F extends CompletableFuture, B, R> extends Ab
     private int currentPartitionIndex = -1;
     private int numCompletedPartitions;
 
+    private Object pendingItem;
+
     private ReadMapOrCacheP(@Nonnull Reader<F, B, R> reader, @Nonnull int[] partitionIds) {
         this.reader = reader;
         this.partitionIds = partitionIds;
@@ -143,13 +145,19 @@ public final class ReadMapOrCacheP<F extends CompletableFuture, B, R> extends Ab
     }
 
     private boolean emitResultSet() {
-        for (; currentBatchPosition < currentBatch.size(); currentBatchPosition++) {
-            Object result = reader.toObject(currentBatch.get(currentBatchPosition));
-            if (result == null) {
+        if (pendingItem != null && !tryEmit(pendingItem)) {
+            return false;
+        }
+        pendingItem = null;
+
+        while (currentBatchPosition < currentBatch.size()) {
+            Object item = reader.toObject(currentBatch.get(currentBatchPosition++));
+            if (item == null) {
                 // element was filtered out by the predicate (?)
                 continue;
             }
-            if (!tryEmit(result)) {
+            if (!tryEmit(item)) {
+                pendingItem = item;
                 return false;
             }
         }

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/ReadMapOrCachePTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/ReadMapOrCachePTest.java
@@ -94,6 +94,23 @@ public class ReadMapOrCachePTest extends SimpleTestInClusterSupport {
                 .expectOutput(expected);
     }
 
+    @Test
+    public void test_whenProjectedToObjectWithNoEquals() {
+        // test for https://github.com/hazelcast/hazelcast-jet/issues/2448
+        IMap<Integer, Object[]> map = instance().getMap(randomMapName());
+        // two values are enough: TestSupport always uses outbox limited to 1 item to drive the processors crazy
+        map.put(0, new Object[0]);
+        map.put(1, new Object[0]);
+
+        TestSupport
+                .verifyProcessor(adaptSupplier(SourceProcessors.readMapP(map.getName())))
+                .jetInstance(instance())
+                .disableSnapshots()
+                .disableProgressAssertion()
+                .outputChecker((expected, actual) -> 2 == actual.size())
+                .expectOutput(emptyList());
+    }
+
     private static <I, O> Projection<I, O> toProjection(FunctionEx<I, O> projectionFn) {
         return (Projection<I, O>) projectionFn::apply;
     }

